docs/yum-debug-restore.1
.\" yum-debug-dump
.TH "yum-debug-restore" "1" "15 December 2011" "James Antill" ""
.SH "NAME"
yum-debug-restore \- replay Yum transactions captured in a debug-dump file
.SH "SYNOPSIS"
\fByum-debug-restore\fP
.SH "DESCRIPTION"
.PP 
\fByum-debug-restore\fP is a program which takes a gzipped file created by
\fByum-debug-dump\fP and acts on the information about installed packages
contained within.
.PP

.SH "GENERAL OPTIONS"
.IP "\fB\-\-output\fP"
Output the commands that would be run to stdout.
.IP "\fB\-\-shell=<file>\fP"
Output the commands that would be run to a file.
.IP "\fB\-\-install\-latest\fP"
Ask yum to install the latest version of the given packages, instead of the
version that was installed in the debug-dump file.
.IP "\fB\-\-ignore\-arch\fP"
Ignore the architecture of the packages, so you can "restore" an i386 debug-dump
on an x86_64 machine.
.IP "\fB\-\-filter\-types=[install,remove,update,downgrade]\fP"
Only perform the given types of commands, so you can filter to just upgrades
and installs.

.SH "FILES"
As yum-debug-restore uses YUM libraries for retrieving all the information, it
relies on YUM configuration for its default values like which repositories
to use. Consult YUM documentation for details:
.PP
.nf 
/etc/yum.conf
/etc/yum/repos.d/
/var/cache/yum/
.fi 

.PP 
.SH "SEE ALSO"
.nf
.I yum-debug-dump (1)
.I yum.conf (5)
http://yum.baseurl.org/
.fi 

.PP 
.SH "AUTHORS"
.nf 
See the Authors file included with this program.
.fi 

.PP 
.SH "BUGS"
There are of course no bugs, but should you find any, you should first
consult the FAQ section on http://yum.baseurl.org/wiki/Faq and if unsuccessful
in finding a resolution contact the mailing list: yum-devel@lists.baseurl.org.
To file a bug use http://bugzilla.redhat.com for Fedora/RHEL/Centos
related bugs and http://yum.baseurl.org/report for all other bugs.

.fi


ChangeLog
2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* ChangeLog: updated ChangeLog

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-03  Shawn Starr <shawn.starr@rogers.com>

	* docs/yum-utils.1: Update yum-utils man page, to add new utils.

2011-07-25  James Antill <james@and.org>

	Merge branch 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils  * 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils: (3 commits)
	typo leaving .conf in ybc.conf.yumvar should just be ybc.yumvar
	...

2011-07-25  James Antill <james@and.org>

	* plugins/aliases/aliases: Add some new aliases, fix chkD.

2011-07-25  Seth Vidal <skvidal@fedoraproject.org>

	* yum-config-manager.py: typo leaving .conf in ybc.conf.yumvar
	should just be ybc.yumvar

2011-07-12  James Antill <james@and.org>

	* plugins/security/security.py: Also auto-apply --security when
	given --sec-severity for other commands.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add repoquery @group completion.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* repoquery.py: Fix "nitmes" typo in previous commit.

2011-07-06  James Antill <james@and.org>

	* repoquery.py: Fix typo

2011-06-29  James Antill <james@and.org>

	* repoquery.py: Allow @foo groups as a packages selector in
	repoquery.

2011-06-25  Ville Skyttä <ville.skytta@iki.fi>

	* plugins/aliases/aliases.py, plugins/security/security.py,
	plugins/versionlock/versionlock.py: Remove some unused local
	variables and imports flagged by pyflakes.

2011-06-24  Ville Skyttä <ville.skytta@iki.fi>

	* debuginfo-install.py, package-cleanup.py, yum-builddep.py,
	yum-complete-transaction.py: Check doUtilBuildTransaction return
	value, exit early on failure.  BZ 716267

2011-06-22  James Antill <james@and.org>

	* plugins/security/security.py: Try to be clever when sorting the
	security severities.

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: After (external) merge is
	complete, give user option to delete other file

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: Support meld as well as vimdiff
	for merging  When using meld, the left side has the "other file"
	i.e. the file with the .rpmnew or .rpmsave extension. The right side
	has the "final file" i.e. the target file into which changes have to
	be merged.  If the rpm action was to save the existing conf file,
	then the left side will be the original conf file (.rpmsave) and the
	right side will be the new rpm-based configuration. The user should
	merge any changes they want to keep from their old .rpmsave from
	left to right.  If the rpm action was to keep the existing conf
	file, then the left side will be the new conf file (.rpmnew) and
	right side will be the current local configuration. The user should
	merge any changes in the new rpm file into their local config from
	left to right.  Ideally these operations would be 3-way diffs
	between the local file and the new rpm config, using the original
	rpm config as the common merge base. However, the original rpm
	config is not available to yum and therefore a 3-way diff is not
	possible. Perhaps a future enhancement would be to use a .orig file,
	if present, as the merge base for meld.  When using vimdiff, the
	order has not changed. The first parameter to vimdiff is the "final:
	file i.e. the target file, and the second parameter is the "other"
	file i.e. the .rpmnew or .rpmsave file.

2011-06-10  James Antill <james@and.org>

	* repodiff.py: Huge speedup for large number of add/remove obsoletes
	processing.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* yumdownloader.py: Stops yumdownloader from attempting to download
	the same package twice if it is available in multiple repos. BZ
	711767

2011-06-10  James Antill <james@and.org>

	* docs/repodiff.1, repodiff.py: Add --compare-arch for the other
	side of the multilib. diff. BZ 710579.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* repodiff.py: Changed archlist set up to handle multilib. BZ 710579


2011-06-01  James Antill <james@and.org>

	* docs/repoquery.1: Add releasever and setopt docs. from yum to
	repoquery. BZ 638542.

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* yum-utils.spec: frelling pyyaml pkg named with weird case

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/puppetverify/puppetverify.conf,
	plugins/puppetverify/puppetverify.py, yum-utils.spec: add
	puppetverify plugin

2011-05-18  Frank Weltersbach <frankwe@live.com>

	* docs/package-cleanup.1: Fix typo in package-cleanup man page. BZ
	705617.

2011-05-17  Seth Vidal <skvidal@fedoraproject.org>

	* repoquery.py:     - fix rh bug: #705179     - fix fmt_queryformat not being happy
	    with new 'output' formats passed       everywhere else.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Allow files/provides to be used in
	yumdownloader.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Exit with code 2, if we can't download the
	packages asked for. BZ 699469.

2011-04-26  James Antill <james@and.org>

	* plugins/auto-update-debuginfo/auto-update-debuginfo.py: try/except
	the open, for the rpmdbv cached data (a-u-d plugin).

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add yum-debug-dump completion.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Mention filename arg in yum-debug-dump usage
	message.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Make yum-debug-dump.py executable for running
	in-place.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Use functions from recent bash-completion if
	available.

2011-04-06  James Antill <james@and.org>

	* repoquery.py, yum-groups-manager.py: Convert

yum-debug-restore.py
#!/usr/bin/python -tt

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
## (c) 2008 Red Hat. Written by skvidal@fedoraproject.org
##                                james@fedoraproject.org

import os
import sys
import gzip
import tempfile

from optparse import OptionParser

import yum
import rpmUtils.miscutils

sections = ['%%%%SYSTEM INFO\n', '%%%%YUM INFO\n',
            '%%%%RPMDB PROBLEMS\n', '%%%%RPMDB\n',
            '%%%%REPOS\n']

def cmd_line():
    parser = OptionParser()
    parser.set_usage("yum-debug-restore [options]")
    parser.add_option("-C", "--cache", action="store_true",
                      help="run from cache only")
    parser.add_option("-c", dest="conffile", help="config file location")
    parser.add_option("--enablerepo", action="append", dest="enablerepos",
                      help="specify additional repoids to query, can be specified multiple times")
    parser.add_option("--disablerepo", action="append", dest="disablerepos",
                      help="specify repoids to disable, can be specified multiple times")                      
    parser.add_option("-y", dest="assumeyes", action="store_true",
                      help="answer yes for all questions")
    parser.add_option("--skip-broken", action="store_true",
                      help="skip packages with depsolving problems")

    parser.add_option("--output", action="store_true",
                      help="output the yum shell commands")
    parser.add_option("--shell", 
                      help="output the yum shell commands to a file")

    parser.add_option("--install-latest", action="store_true",
                      help="install the latest instead of specific versions")
    parser.add_option("--ignore-arch", action="store_true",
                      help="ignore arch of packages, so you can dump on .i386 and restore on .x86_64")

    parser.add_option("--filter-types", 
                      help="Limit to: install, remove, update, downgrade")

    (opts, args) = parser.parse_args()
    if not args:
        parser.print_usage()
        sys.exit(1)
    return (opts, args)

class OtherRpmDB:

    def __init__(self, fn):
        self.pkgtups = []

        if fn.endswith(".gz"):
            fo = gzip.GzipFile(fn)
        else:
            fo = open(fn)

        if fo.readline() != 'yum-debug-dump version 1\n':
            print >>sys.stderr, "Bad yum debug file:", fn
            sys.exit(1)

        skip = sections[:-1]
        for line in fo:
            if skip: # Header stuff
                if line == skip[0]:
                    skip.pop(0)
                continue

            if not line or line[0] != ' ':
                break

            pkgtup = rpmUtils.miscutils.splitFilename(line.strip())
            n,v,r,e,a = pkgtup # grrr...
            pkgtup = (n,a,e,v,r)
            self.pkgtups.append(pkgtup)

def naevr2str(n,a,e,v,r):
    if a is None: # Assume epoch doesn't change without release changing
        return "%s-%s-%s" % (n,v,r)
    if e in (None, '', '0'):
        return "%s-%s-%s.%s" % (n,v,r,a)
    return "%s-%s:%s-%s.%s" % (n,e,v,r,a)
def pkgtup2str(pkgtup):
    n,a,e,v,r = pkgtup
    return naevr2str(n,a,e,v,r)

def pkg_data2list(yb, opkgtups, opkgmaps, install_latest, ignore_arch):
    ret = []
    npkgtups = set()
    npkgmaps = {}
    installonly = set(yb.conf.installonlypkgs)
    for po in sorted(yb.rpmdb.returnPackages()):
        arch = po.arch
        if ignore_arch:
            arch = None
        if False: pass
        elif po.name in installonly:
            if not po.pkgtup in opkgtups:
                ret.append(("remove", pkgtup2str(po.pkgtup)))
        elif (po.name, arch) not in opkgmaps:
            ret.append(("remove", str(po)))
        elif po.pkgtup not in opkgtups:
            n,a,e,v,r = opkgmaps[(po.name, arch)]
            pinstEVR = yum.packages.PackageEVR(e, v, r)
            if po.EVR == pinstEVR:
                assert ignore_arch and po.arch != a
            elif po.EVR < pinstEVR:
                ret.append(("upgrade",   naevr2str(n,arch,e,v,r)))
            else:
                ret.append(("downgrade", naevr2str(n,arch,e,v,r)))
        npkgtups.add(po.pkgtup)
        npkgmaps[(po.name, po.arch)] = po
        if ignore_arch:
            npkgmaps[(po.name, None)] = po

    for name, arch in sorted(opkgmaps):
        if name in installonly:
            continue # done separately
        if ignore_arch and arch is not None:
            continue
        if (name, arch) in npkgmaps:
            continue
        if install_latest and ignore_arch:
            ret.append(("install", name))
        elif install_latest:
            ret.append(("install", "%s.%s" % (name, arch)))
        else:
            ret.append(("install", pkgtup2str(opkgmaps[(name, arch)])))
    for pkgtup in opkgtups:
        if pkgtup[0] in installonly and not pkgtup in npkgtups:
            ret.append(("install", pkgtup2str(pkgtup)))
    return ret

def main():
    (opts, args) = cmd_line()
    yb = yum.YumBase()
    yb.preconf.init_plugins = True
    if opts.conffile:
        yb.preconf.fn = opts.conffile

    yb.conf
    if opts.cache:
        yb.conf.cache = True

    if opts.disablerepos:
        for repo_match in opts.disablerepos:
            for repo in yb.repos.findRepos(repo_match):
                repo.disable()

    if opts.enablerepos:    
        for repo_match in opts.enablerepos:
            for repo in yb.repos.findRepos(repo_match):
                repo.enable()

    xtra_args = []
    if opts.skip_broken:
        xtra_args.append('--skip-broken')

    if opts.assumeyes:
        xtra_args.append('-y')

    fn = args[0]
    print "Reading from: %s" % fn
    orpmdb = OtherRpmDB(fn)

    opkgmaps = {}
    for pkgtup in orpmdb.pkgtups:
        opkgmaps[(pkgtup[0], pkgtup[1])] = pkgtup
        if opts.ignore_arch:
            n,a,e,v,r = pkgtup
            opkgmaps[(pkgtup[0], None)] = n,None,e,v,r

    if opts.output:
        fo = sys.stdout
    elif opts.shell:
        try:
            fo = open(opts.shell, "wb")
        except OSError, e:
            print >>sys.stderr, "open(%s): %s" % (opts.shell, e)
            sys.exit(1)
    else:
        fo = tempfile.NamedTemporaryFile()

    fT = None
    if opts.filter_types:
        fT = set(opts.filter_types.replace(",", " ").split())

    counts = {}
    for T, pkg in pkg_data2list(yb, set(orpmdb.pkgtups), opkgmaps,
                                opts.install_latest, opts.ignore_arch):
        if fT is not None and T not in fT:
            continue
        counts[T] = counts.get(T, 0) + 1
        try:
            print >>fo, "%-9s %s" % (T, pkg)
        except IOError:
            if opts.output: # mainly due to |
                sys.exit(1)
            raise

    if opts.output:
        sys.exit(0)

    print >>fo, "run"
    fo.flush()

    if opts.shell:
        if counts:
            print "Statistics:"
        for T in sorted(counts):
            print "    %9s %6u" % (T, counts[T])
        print "Done"
        sys.exit(0)

    # Want to do the transaction, hacky method
    xtra_args.append('--setopt=installonly_limit=0')
    os.system("yum shell %s %s" % (" ".join(xtra_args), fo.name))

if __name__ == "__main__":
    main()


